### PR TITLE
Prevent Icon from saving when the upload fails

### LINKED ISSUE #11980: Icon is saved even when the upload fails
### Context
If you are trying to upload an icon, and it fails (because of the size validation for instance), the icon is saved in `Your uploads`. This happens for single or multiple uploads:

icon-upload-error

### Steps to Reproduce

1. Open the icon dialog. 
2. Select a source that doesn't fit into the expected file. For instance, the max size is 1024x1024. Choose a file bigger than that.
3. Once it fails, go to `Your uploads` tab.

### Current Result
The icon is saved but the image source is obviously broken.

### Expected result
The icon is not saved.

### PR DESCRIPTION
**Pull request by kutyel**
Fixes #11980 

carto

### CODE REVIEW

**Comment by Cartofante** [comment]
Can one of the admins verify this patch?

**Comment by Cartofante** [comment]
Can one of the admins verify this patch?

**Comment by nobuti** [comment]
Nice. Could you include a test?

**Comment by kutyel** [comment]
Sure thing! Here you go 😉 👍

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js [hunk 1]
```diff
@@ -235,6 +235,14 @@ describe('components/form-components/editors/fill/input-color/assets-view', func
       expect(this.view._stateModel.get('error_message')).toBe('hi, I am an error');
       expect(this.view._stateModel.get('status')).toBe('error');
     });
+
+    it('should preserve the number of assets when upload fails', function () {
+      var previousAssetsCount = this.view._userAssetCollection.length;
+
+      mockCreate.call(this, 'error');
```
Comment (nobuti) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js]: Correct me if I'm wrong but I think `mockCreate` doesn't do anything on its own, you need to "upload" an image to force a call to the create method of the collection, then the error callback.

Comment (kutyel) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js]: You are right, I played along with the tests and couldn't find an actual way to test this behaviour, for example I tried spying on the `this.view._userAssetCollection.remove' method and it didn't work either. Would really appreciate some hints on how to approach this unit test.

Comment (nobuti) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js]: I think the approach is good, you only have to find the way to simulate an upload. If you connect the dots, from the moment the user click on the upload button, you'll find the answer.

File: lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js [hunk 2]
```diff
@@ -235,6 +235,16 @@ describe('components/form-components/editors/fill/input-color/assets-view', func
...
+
+      this.view.$('.js-fileInput').trigger('change');
```
Comment (kutyel) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js]: @nobuti what about this? 😅

Comment (nobuti) [inline lib/assets/core/test/spec/cartodb3/components/form-components/editors/fill/input-color/asset-picker/assets-view.spec.js]: Nice!

**Comment by xavijam** [comment]
Hey @kutyel, thanks for this 🤗 , closing it.

**Pull request closed without merging**